py-esmpy: new package

angus-g opened the pull request:
This is the Python interface to ESMF. I'm not sure whether this is the best approach, or if it should be added as a variant to ESMF? This seemed like the most straightforward method, given it's a simple Python package, bundled with the ESMF repository.

This depends on py-setuptools-git-versioning #36123

angus-g committed:
py-esmpy: new package

This is the Python interface to ESMF

tldahlgren commented on var/spack/repos/builtin/packages/py-esmpy/package.py:
```diff
@@ -0,0 +1,34 @@
+    depends_on("py-setuptools", type="build")
+    depends_on("py-setuptools-git-versioning", type="build")
+
+    build_directory = "src/addon/esmpy"
```
Portability would suggest the following change:

```suggestion
    build_directory = join_path("src", "addon", "esmpy")
```

tldahlgren reviewed (commented):
Can confirm the branch and all tags EXCEPT `8.1.1` and `8.0.1` exist.

angus-g commented:
> Can confirm the branch and all tags EXCEPT `8.1.1` and `8.0.1` exist.

Ah yes, good catch! In fact, the build directory was moved from `src/addon/ESMPy` in 8.3.1 and before to `src/addon/esmpy` in 8.4.0 and after. This version split also corresponds to the `py-setuptools-git-versioning` dependency: I assume the older versions could be built from a tarball, and then we don't need to refer to them by tag.

tldahlgren commented:
> > Can confirm the branch and all tags EXCEPT `8.1.1` and `8.0.1` exist.
> 
> Ah yes, good catch! In fact, the build directory was moved from `src/addon/ESMPy` in 8.3.1 and before to `src/addon/esmpy` in 8.4.0 and after. This version split also corresponds to the `py-setuptools-git-versioning` dependency: I assume the older versions could be built from a tarball, and then we don't need to refer to them by tag.
> 
> How would spack deal with a split between offering versions as git tags vs. tarballs? Does it preference one over the other?

I would be surprised if there isn't at a least on package that does this but I'd have to check to be sure.

adamjstewart commented on var/spack/repos/builtin/packages/py-esmpy/package.py:
```diff
@@ -0,0 +1,34 @@
+
+    # because it relies on py-setuptools-git-versioning, we need to install from git
+    version("develop", branch="develop")
+    version("8.4.0", tag="v8.4.0")
+    version("8.3.1", tag="v8.3.1")
+    version("8.3.0", tag="v8.3.0")
+    version("8.2.0", tag="v8.2.0")
+    version("8.1.1", tag="v8.1.1")
+    version("8.0.1", tag="v8.0.1")
+    version("8.0.0", tag="v8.0.0")
```
URLs are greatly preferred over git checkouts whenever possible for security reasons. Git tags can be moved after the fact, and can't be checksummed. But if git is required to use setuptools-git-versioning, then I guess there isn't much you can do.

angus-g replied:
I'll have a bit of a deeper dig to see if I can make the tarballs build. If they don't, it seems like an upstream bug that should be fixed anyway!

Out of curiosity, what are the security reasons? Would it be safer to also specify the commit hash associated with the tags? That seems to me to be fairly similar to a tarball + sha256.

adamjstewart replied:
Tarball + sha256 is safest because it's easy to verify the integrity of the tarball before extracting it. Commit hash isn't as good because when Spack downloads something from git it can verify it the first time, but it can't be verified after it's been archived in a tarball. We strip the .git directory for some reason, so there's no way to verify it even after extracting. Tag is the least safe since it can't be verified during the download or during usage. @tgamblin know more about this than I do though.

adamjstewart commented on var/spack/repos/builtin/packages/py-esmpy/package.py:
```diff
@@ -0,0 +1,34 @@
+    for v in ["develop", "8.4.0", "8.3.1", "8.3.0", "8.2.0", "8.1.1", "8.0.1", "8.0.0"]:
+        depends_on(f"esmf@{v}", when=f"@{v}", type=("build", "run"))
+
+    depends_on("py-setuptools", type="build")
```
```suggestion
    depends_on("py-setuptools@41:", type="build")
```

adamjstewart commented on var/spack/repos/builtin/packages/py-esmpy/package.py:
```diff
@@ -0,0 +1,34 @@
+        depends_on(f"esmf@{v}", when=f"@{v}", type=("build", "run"))
+
+    depends_on("py-setuptools", type="build")
+    depends_on("py-setuptools-git-versioning", type="build")
```
Missing dependency on numpy

adamjstewart commented:
If you're having issues with older versions, you could just only include recent working versions if you want to keep the package simple.

angus-g committed:
Specify required py-setuptools version

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

angus-g committed:
Convert to tarball-based builds

Turns out py-setuptools-git-versioning doesn't necessarily need to run
from a git repository, so we should be able to use tarballs rather
than git checkouts.

adamjstewart commented on var/spack/repos/builtin/packages/py-esmpy/package.py:
```diff
@@ -0,0 +1,54 @@
+    """ESMPy is a Python interface to the Earth System Modeling Framework (ESMF)"""
+
+    homepage = "https://earthsystemmodeling.org/"
+    git = "https://github.com/esmf-org/esmf.git"
```
Can you add a url attribute too? It's used by `spack versions` and `spack checksum`.

adamjstewart commented on var/spack/repos/builtin/packages/py-esmpy/package.py:
```diff
@@ -0,0 +1,54 @@
+    homepage = "https://earthsystemmodeling.org/"
+    git = "https://github.com/esmf-org/esmf.git"
+
+    # because it relies on py-setuptools-git-versioning, we need to install from git
```
This comment is seemingly no longer relevant

adamjstewart commented on var/spack/repos/builtin/packages/py-esmpy/package.py:
```diff
@@ -0,0 +1,54 @@
+
+    depends_on("py-setuptools@41:", type="build")
+    depends_on("py-build", when="@:8.3.1", type="build")
+    depends_on("py-setuptools-git-versioning", when="@8.4.0:", type="build")
```
```suggestion
    depends_on("py-setuptools-git-versioning", when="@8.4.0: ^python@3.10:", type="build")
```

adamjstewart commented on var/spack/repos/builtin/packages/py-esmpy/package.py:
```diff
@@ -0,0 +1,54 @@
+    depends_on("py-setuptools@41:", type="build")
+    depends_on("py-build", when="@:8.3.1", type="build")
+    depends_on("py-setuptools-git-versioning", when="@8.4.0:", type="build")
+    depends_on("py-numpy", type="run")
```
```suggestion
    depends_on("py-numpy", type=("build", "run"))
```

adamjstewart replied:
Also missing dep on importlib-metadata and pytest-json-report

angus-g replied:
Is pytest-json-report needed, since it's a testing-optional dependency? Not sure what the testing protocol is for spack. Happy to add it as a package and then a dependency though!

adamjstewart replied:
I'm sure it's only used for testing, but it's listed in `pyproject.toml` as a required dependency. Although we currently disable checks for dependencies during the install, that's something I would like to be able to fix someday. So it's best to add it as a dependency like it's documented, or to patch it and remove it. If you do patch it, I would open a PR with esmf to fix this.

angus-g replied:
Ah of course, I was looking at the git version where it had been moved to an optional dependency already. I've just added that into this PR. I can move history around to get the commits in the right order and have one commit for each of py-esmpy and py-pytest-json-report if needed.

angus-g committed:
Add URL attribute to py-esmpy

angus-g committed:
Fix py-esmpy dependencies

angus-g committed:
py-pytest-json-report: new package

angus-g committed:
Add py-pytest-json-report as py-esmpy dependency

adamjstewart commented on var/spack/repos/builtin/packages/py-pytest-json-report/package.py:
```diff
@@ -0,0 +1,20 @@
+
+    version("1.5.0", sha256="2dde3c647851a19b5f3700729e8310a6e66efb2077d674f27ddea3d34dc615de")
+
+    depends_on("py-pytest@3.8.0:")
+    depends_on("py-pytest-metadata")
```
```suggestion
    depends_on("py-setuptools", type="build")
    depends_on("py-pytest@3.8:", type=("build", "run"))
    depends_on("py-pytest-metadata", type=("build", "run"))
```

angus-g replied:
Oops, completely forgot how to make a Python package for a second there 😅

angus-g committed:
Fix py-pytest-json-report dependencies

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

adamjstewart commented on var/spack/repos/builtin/packages/py-esmpy/package.py:
```diff
@@ -0,0 +1,56 @@
+    depends_on("py-build", when="@:8.3.1", type="build")
+    depends_on("py-setuptools-git-versioning", when="@8.4.0: ^python@3.10:", type="build")
+    depends_on("py-numpy", type=("build", "run"))
+    depends_on("py-importlib-metadata", when="@8.4.0: ^python@:3.7")
+    depends_on("py-pytest-json-report")
```
```suggestion
    depends_on("py-importlib-metadata", when="@8.4: ^python@:3.7", type=("build", "run"))
    depends_on("py-pytest-json-report", when="@8.4:", type=("build", "run"))
```

adamjstewart commented on var/spack/repos/builtin/packages/py-esmpy/package.py:
```diff
@@ -0,0 +1,56 @@
+        depends_on(f"esmf@{v}", when=f"@{v}", type=("build", "run"))
+
+    depends_on("py-setuptools@41:", type="build")
+    depends_on("py-build", when="@:8.3.1", type="build")
```
Can't remember if we discussed this, but I don't think build is required. We use pip to install.

angus-g replied:
I think I added this because the older versions (those without `pyproject.toml`) wouldn't install through pip/setuptools.

adamjstewart replied:
What was the error?

climbfuji commented:
Just came across this stale PR. NOte that there is a newer PR https://github.com/spack/spack/pull/45504 that adds a Python extension to ESMF, which I think is the right way to do this? @chrismarsh FYI
